lib/utils.nix

```diff
@@ -4,7 +4,15 @@
   _nixvimTests,
 }:
 with lib;
-{
+rec {
+  # Whether a string contains something other than whitespace
+  hasContent = str: builtins.match "[[:space:]]*" str == null;
+
+  # Concatenate a list of strings with a newline between each element
+  # skipping strings containing only whitespace characters
+  concatNonEmptyStringsNewline =
```

Comment:
That was a hard one. `concatNonEmptyLines` was one of the choices, but it's similar to `concatLines` from nixpkgs library and their behavior is different. concatLines is for concatenating one-line-strings, that's why it appends newlines to every string. This function does a different thing. I named it after concatStringsSep with a Sep suffix changed to the actual separator, since my function doesn't separator as an argument.

But since you suggested `concatNonEmptyLines` as a best choice, so be it. I don't like my choice anyway.

Reply:
~~[This looks like](https://noogle.dev/f/lib/concatLines) the same behaviour to me:~~

```nix
concatLines [ "foo" "bar" ]
=> "foo\nbar\n"
```

`concatLines` is _almost_ the same as `concatStringsSep "\n"`.

Other than the trailing newline, which is only added by `concatLines`, incidentally that's probably what we actually _want_ in this case?

Maybe the implementation should actually be `lines: concatLines (filter hasContent lines)`?

Reply:
> Maybe the implementation should actually be `lines: concatLines (filter hasContent lines)`?

No, actually I don't want a trailing newline. The reason for this is that most of the time config parts are done like this:

```nix
{
  var = ''
    text
  ''
}
```
This adds a trailing newline by itself. With `concatLines` it'll be two trailing newlines. In my config I just make sure, that all my parts have trailing newlines by itself (parts from files also have trailing newlines, because I don't use `: set nofixendofline`. But the resulting concatenated result wouldn't be right if some strings are assigned with `var = "text";`. To fix this case I usually use a map with first stripping trailing newlines, than adding exactly one.

Reply:
I'm not sure I agree, since many of the places changed in _this_ PR are going from either using the pattern you describe or using `lib.concatLines` to now using `helpers.concatNonEmptyLines` _without_ a trailing newline.

Since the trailing newline is only added when there is content, I'm not sure I see the issue? Maybe I'm missing your point?

We could provide a `helpers.concatNonEmptyLines'` that takes an additional "trailing" boolean param, but this is a little over-engineered IMO.

Reply:
> I'm not sure I see the issue

Since resulted config is passed through stylua, it's not a big deal. As I mentioned in comments several days ago, I often nit-picking those little things, so that my generated config would be nearly perfect. There is another case I didn't touch here, that I handled in my config: when doing `wrapDo` you actually need to indent every line (except empty lines). Because it's not handled now I intentionally wasn't indenting a first line of the string in `wrapVimscriptForLua`.

> since many of the places changed in _this_ PR are going from either using the pattern you describe or using `lib.concatLines` to now using `helpers.concatNonEmptyLines` _without_ a trailing newline.

What are many places? There is only one place where I changed `concatLines` to `helpers.concatNonEmptyLines`. It's in the opts module. And my version doesn't have problems, because those strings, that are concatenated by `concatNonEmptyLines` all has trailing newlines by itself. But old version has two trailing newlines (one from indented string and one from cancatLines).

For best results all strings concatenated by `concatNonEmptyLines` should have traling newlines. We can handle strings not having trailing newlines by ensuring all parts are have them. `removeSuffix "\n" + "\n"` should do it.

Reply:
A more general function that takes `sep` as an argument probably shouldn't be normalising line endings, since such functions are not always used on _lines_.

BTW the mapping function `(s: lib.removeSuffix "\n" s + "\n")` will always insert trailing newlines, like `lib.concatLines` does. It's just less likely to insert additional newlines if it is called with "dirty" input.

I think we should avoid our functions being too magic in unexpected ways. IMO the simplest implementation is to match upstream functionality, in this case just use `lib.concatLines` but filter the list for non-empty lines.

If we have an issue of passing in unintended `\n` in the input lines then we should address that in the calling code (could be another PR).

> It is similar to `concatLines` now, but with separator.

I don't see why we would ever need a different separator to `\n` if we know we're concatenating lines?

IMO this is getting overly complex for the sake of additional newlines that a) aren't that important for _this_ PR and b) should really be addressed in the calling code instead of the helper function.

Reply:
> Can you be more specific? What I need to do?

I think we should have:

```nix
concatNonEmptyLines = lines: concatLines (filter hasContent lines);
```

> Because I don't see a problem with my function. I never use `concatLines` myself, because the behavior I want almost always is `concatStringsSep "\n"` with a filter for empty strings.

If you really want, we could also have:

```nix
concatNonEmptyLines' =
  trailing: lines:
  let
    lines' = filter hasContent lines;
  in
  optionalString (lines' != [ ]) (
    concatStringsSep "\n" lines'
    + optionalString trailing "\n"
  );

concatNonEmptyLines = concatNonEmptyLines' true;
```

Depending on how often we really need that.

Reply:
> b) should really be addressed in the calling code instead of the helper function.

So if I want it to be addressed _in the calling code_, do you recommend me to use this, righ?

```nix
    content =
      if config.type == "lua" then
        # Lua
        helpers.concatNonEmptyLines [
          config.extraConfigLuaPre
          (helpers.wrapVimscriptForLua config.extraConfigVim)
          config.extraConfigLua
          (lib.removeSuffix "\n" config.extraConfigLuaPost)
        ]
      else
        # Vimscript
        helpers.concatNonEmptyLines [
          (helpers.wrapLuaForVimscript config.extraConfigLuaPre)
          config.extraConfigVim
          (lib.removeSuffix "\n" (
            helpers.wrapLuaForVimscript (
              helpers.concatNonEmptyLines [
                config.extraConfigLua
                (lib.removeSuffix "\n" config.extraConfigLuaPost)
              ]
            )
          ))
        ];
```

Reply:
Anyway, you are a maintainer here, so I'll change as you suggested even though your suggested function definitely not want I need. But anyway, let's it have extra newlines. I don't have enough energy to argue, better to finish it somehow, because I guess we won't agree on this one.

Reply:
If `extraConfigLuaPost` is likely to have trailing newline(s) then so is `extraConfigLua`, `extraConfigVim` and `extraConfigLuaPre`.

If we're getting that nit-picky, then `removeSuffix "\n"` isn't a robust solution, since there could be any number of trailing newlines hypothetically.

IMHO this isn't in the scope of this PR and we aren't making things any worse than the existing implementation by going with the simple option for now.

If you really think it's important to normalise the ending of those options, maybe we can do that in the options' `apply` functions? But again, I think we should keep that out of _this_ PR.

Reply:
> I'll change as you suggested even though your suggested function definitely not what I need. But anyway, let's it have extra newlines. I don't have enough energy to argue

Sorry if that got unnecessarily heated. That wasn't my intention! :sweat_smile: 

I'm not saying we definitely shouldn't deal with redundant newlines, I just don't want to deal with that additional complexity in this PR since it primarily focuses on keeping the output empty by default.

I'm also not saying we definitely should fix them, but we can discuss potential solutions in another issue/discussion/PR/matrix.

Reply:
> If extraConfigLuaPost is likely to have trailing newline(s) then so is extraConfigLua, extraConfigVim and extraConfigLuaPre.

But my problem is with with `concatNonEmptyLines` producing unnecessarily newline at the end, so I don't need to strip other variables. They will be concatenated properly. Only the last one shouldn't have newline to count the one added by concatLines. So my complain is that you take my function that get the job done (because my guess is that at least 90% strings already have trailing newlines due to using indented string style) and turn in the function that doesn't.

> If we're getting that nit-picky, then `removeSuffix "\n"` isn't a robust solution, since there could be any number of trailing newlines hypothetically.

That's right, but caller may want it to have some trailing newlines, that's why I usually strip only one. Usecase for this is that both strings from here will be similar:

```nix
{
  a = "txt;
  b = ''
    txt
  ''
}
```

> IMHO this isn't in the scope of this PR and we aren't making things any worse than the existing implementation by going with the simple option for now.

Originally not, but when you suggested improving, I take it to my heart. So with my second commit I wanted to not only produce _whitespace only_ config, but clean up config generation. Anyway, I already gave in to do it your way. I guess comments are refreshed too late, because we often reply to the penultimate comments.